### PR TITLE
Refactor place hold results

### PR DESCRIPTION
**Pull request by banukutlu**
Issue #151 

## Note
✔ #147 needs to be merged before this

### CODE REVIEW

File: app/services/place_hold_form/builder.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ def initialize(catkey:, user_token:, client:)
   end
 
   def generate
-    bib_info = Bib.new(parsed_symphony_response(:get_bib_info, @catkey, @user_token))
+    bib_info = Bib.new(SymphonyClientParser::parsed_response(@client, :get_bib_info, @catkey, @user_token))
```
Comment (cdmo) [inline app/services/place_hold_form/builder.rb]: Thanks for updating this

File: app/controllers/holds_controller.rb [hunk 2]
```diff
@@ -72,18 +72,14 @@ def result
...
-    session.delete(:place_hold_catkey)
 
-    raw_place_hold_results = session[:place_hold_results]
-    session.delete(:place_hold_results)
+    results_builder = PlaceHoldResults::Builder.new(user_token: current_user.session_token,
+                                                    client: symphony_client,
+                                                    place_hold_results: session[:place_hold_results])
+    @place_hold_results = results_builder.generate
 
-    @place_hold_results = raw_place_hold_results.each do |status, results|
```
Comment (cdmo) [inline app/controllers/holds_controller.rb]: really happy to see this code and related methods moved out of the controller

File: app/services/symphony_client.rb [hunk 3]
```diff
@@ -166,6 +166,8 @@ def error_code(response)
       return if response.status.ok?
 
       JSON.parse(response.body).dig('messageList')[0].dig('code')
+    rescue JSON::ParserError
+      nil
```
Comment (cdmo) [inline app/services/symphony_client.rb]: I want to make sure I understand this bit: When JSON tries to parse the response body and then dig for code, if it hits a bad formatting problem or whatever in the JSON itself, it will raise up and set error_code to nil?

File: app/services/symphony_client.rb [hunk 4]
```diff
@@ -166,6 +166,8 @@ def error_code(response)
       return if response.status.ok?
 
       JSON.parse(response.body).dig('messageList')[0].dig('code')
```
Comment (cdmo) [inline app/services/symphony_client.rb]: Not part of this PR, but what do you think of:

```suggestion
      JSON.parse(response.body).dig('messageList').first.dig('code')
```

I think it's easier to read, more ruby-ish

File: app/services/symphony_client.rb (hunk 3 above)
Comment (banukutlu) [reply app/services/symphony_client.rb]: yup, will return nil for any response without the expected messageList and code fields

Comment (cdmo) [reply app/services/symphony_client.rb]: this would only catch errors during `parse` not during `dig` or `first`... right?

File: spec/services/place_hold_results/builder_spec.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
```
Comment (cdmo) [inline spec/services/place_hold_results/builder_spec.rb]: Should we add a spec to test that the loop is being executed too? Like you did for #147?

File: spec/services/place_hold_results/builder_spec.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+  describe '#generate' do
+    it 'will pass along processed placed holds results' do
+      expect(processed_results[:success].first).to include(hold_key: 'a_hold_key')
+        .and include(barcode: 'holdable_barcode')
+        .and include(:placed_hold)
+    end
+
+    it 'returns Hold objects for placed holds' do
+      expect(processed_results[:success].first[:placed_hold]).to be_a Hold
```
Comment (cdmo) [inline spec/services/place_hold_results/builder_spec.rb]: This spec and the one that starts on 49 use type checking. I kinda think it's good to avoid type checking, and maybe instead check for something unique about Hold. For example:

`expect(processed_results[:success].first[:placed_hold].pickup_library_human).to eq 'Pattee Library and Paterno Library Stacks'`

I think this makes the code more flexible, less brittle. If we ever renamed `Hold` this spec would start failing as it is, but if we just test against a unique characteristic of Hold (like the human pickup library name) it wouldn't.

For `Item` we could use `barcode` for example.

What do you think?

File: app/services/symphony_client.rb (hunk 3 above)
Comment (banukutlu) [reply app/services/symphony_client.rb]: I think so yes

Comment (cdmo) [reply app/services/symphony_client.rb]: I wonder if it'd `raise` this if there was a problem, like if dig or first hit a fatal. Anyway, this is good.

File: spec/services/place_hold_results/builder_spec.rb (hunk 5 above)
Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: if you mean the records in use workaround, I am thinking that we test it with symphony_client_spec, and results just makes calls to symphony_client methods so kinda leaning towards to no need to that here again but would like to hear your reasoning, I might be missing sth.

File: spec/services/place_hold_results/builder_spec.rb (hunk 6 above)
Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: I see your point, just trying to find uniqueness for Item 🤔 I am not sure if barcode is unique to Item only.

File: spec/services/place_hold_results/builder_spec.rb (hunk 5 above)
Comment (cdmo) [reply spec/services/place_hold_results/builder_spec.rb]: So the workaround in #147 relies on `SymphonyClient::authenticated_request` to delay if `records_in_use?`. This one is different. It is checking for a malformed response from the WS. Then it will delay here, in this class (not in SymphonyClient::autheticated_request`. We are not testing a scenario where this results builder encounters malformed response to looking up a hold. 

Strange to me that CC is saying that the line inside the `while` loops is covered, so maybe I'm mis-reading.

File: spec/services/place_hold_results/builder_spec.rb (hunk 6 above)
Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: ok, makes sense, I changed the Hold spec as you suggested and deleted the Item one since there is already another example checking the barcode. So, it should be good now

File: spec/services/place_hold_results/builder_spec.rb (hunk 5 above)
Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: ahhhh ok yeah, the spec for that is still at holds_controller_spec https://github.com/psu-libraries/myaccount/blob/151-refactor-results/spec/controllers/holds_controller_spec.rb#L367, I will try to move it over

File: spec/services/place_hold_results/builder_spec.rb (hunk 6 above)
Comment (cdmo) [reply spec/services/place_hold_results/builder_spec.rb]: Yup. I think this is good too. Thanks.

File: spec/services/place_hold_results/builder_spec.rb (hunk 5 above)
Comment (cdmo) [reply spec/services/place_hold_results/builder_spec.rb]: Ah that's why some confusion. Hope it's an easy port over. Thanks!

Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: yes sorry, I apparently forgot to update these, working on it now :)

Comment (cdmo) [reply spec/services/place_hold_results/builder_spec.rb]: No worries!!

Comment (banukutlu) [reply spec/services/place_hold_results/builder_spec.rb]: ok, pushed a commit for this change, hope it looks right
